src/EditorFeatures/CSharpTest/CodeActions/SyncNamespace/CSharpSyncNamespaceTestsBase.cs

```diff
@@ -0,0 +1,205 @@
+        }
+
+        protected async Task TestMoveFileToMatchNamespace(string initialMarkup, List<string[]> expectedFolders = null)
+        {
+            var testOptions = new TestParameters();
+            using (var workspace = CreateWorkspaceFromOptions(initialMarkup, testOptions))
+            {
+                if (expectedFolders?.Count > 0)
+                {
+                    var expectedFolderPaths = expectedFolders.Select(f => string.Join(PathUtilities.DirectorySeparatorStr, f));
+
+                    var oldDocument = workspace.Documents[0];
+                    var oldDocumentId = oldDocument.Id;
+                    var expectedText = workspace.Documents[0].TextBuffer.CurrentSnapshot.GetText();
+
+                    // a new document with the same text as old document is added.
+                    var allResults = await TestOperationAsync(testOptions, workspace, expectedText);
+
+                    var actualFolderPaths = new HashSet<string>();
+                    foreach (var result in allResults)
+                    {
+                        // the original source document does not exist in the new solution.
+                        var oldSolution = result.Item1;
+                        var newSolution = result.Item2;
+
+                        Assert.Null(newSolution.GetDocument(oldDocumentId));
+
+                        var newDocument = GetDocumentToVerify(expectedChangedDocumentId: null, oldSolution, newSolution);
+                        actualFolderPaths.Add(string.Join(PathUtilities.DirectorySeparatorStr, newDocument.Folders));
+                    }
+
+                    Assert.True(expectedFolderPaths.Count() == actualFolderPaths.Count, "Number of available \"Move file\" actions are not equal.");
+                    foreach (var expected in expectedFolderPaths)
+                    {
+                        Assert.True(actualFolderPaths.Contains(expected));
+                    }
+                }
+                else
+                {
+                    var (actions, _) = await GetCodeActionsAsync(workspace, testOptions);
+                    if (actions.Length > 0)
+                    {
+                        var renameFileAction = actions.Any(action => action is CSharpSyncNamespaceService.MoveFileCodeAction);
+                        Assert.False(renameFileAction, "Rename File to match type code action was not expected, but shows up.");
+                    }
+                }
+            }
+
+            async Task<List<Tuple<Solution, Solution>>> TestOperationAsync(
+                TestParameters parameters,
+                TestWorkspace workspace,
+                string expectedCode)
+            {
+                var results = new List<Tuple<Solution, Solution>>();
+
+                var (actions, _) = await GetCodeActionsAsync(workspace, parameters);
+                var moveFileActions = actions.Where(a => a is CSharpSyncNamespaceService.MoveFileCodeAction);
```

Comment:
yikes.  that seems super leaky.  why do we need this?

Reply:
Multiple options will be provided by the refactoring, this is an easier way to get the code action what we want to test.

Reply:
That's already possible with our test harness, which allows passing an index along to mention which of the options you want.  This also ensures things like ordering, and expresses the test relative to the experience we expect the user to have.